File: server/internal/game/lobby.go
// Package game is the Undercover state machine: one Lobby, mutated by the
// methods below, with no I/O. Every rule here is a port of the Firestore
// transactions the Flutter app used to run client-side, so the JSON field
// names deliberately match the old document.
package game

import (
	"maps"
	"math/rand/v2"
	"slices"
	"strings"
	"time"
)

type Phase string

const (
	PhaseLobby     Phase = "lobby"
	PhaseRevealing Phase = "revealingRoles"
	PhasePlaying   Phase = "playing"
	PhaseGameOver  Phase = "gameOver"
)

const (
	RoleCivilian   = "Civilian"
	RoleUndercover = "Undercover"
	RoleSpectator  = "Spectator"

	WinnerCivilians  = "Civilians"
	WinnerUndercover = "Undercover"

	// SkipVote is the vote value meaning "no elimination this round".
	SkipVote    = "skip"
	DefaultIcon = "assets/default_icon.jpg"

	MinPlayers = 3
)

// Error is a rejected command. Code is what the client switches on.
type Error struct {
	Code    string
	Message string
}

func (e *Error) Error() string { return e.Code + ": " + e.Message }

var (
	ErrNotFound   = &Error{"notFound", "lobby does not exist"}
	ErrExists     = &Error{"exists", "lobby ID already exists"}
	ErrInProgress = &Error{"inProgress", "game in progress"}
	ErrNameTaken  = &Error{"nameTaken", "name already taken"}
	ErrNotHost    = &Error{"notHost", "only the host can do that"}
	ErrExpired    = &Error{"expired", "session expired"}
	ErrInvalid    = &Error{"invalid", "invalid command"}
)

func invalid(msg string) error { return &Error{"invalid", msg} }

// Lobby is persisted whole as JSON. Collections are never nil so the JSON
// the client sees always has [] / {} rather than null.
type Lobby struct {
	ID                 string            `json:"id"`
	Host               string            `json:"host"`
	Players            []string          `json:"players"`
	GameStarted        bool              `json:"gameStarted"`
	GamePhase          Phase             `json:"gamePhase"`
	CreatedAt          int64             `json:"createdAt"` // unix ms
	Roles              map[string]string `json:"roles"`
	SelectedWord       string            `json:"selectedWord"`
	SelectedIcon       string            `json:"selectedIcon"`
	SelectedIsChampion bool              `json:"selectedIsChampion"`
	AlivePlayers       []string          `json:"alivePlayers"`
	RoundOrder         []string          `json:"roundOrder"`
	CurrentPlayerIndex int               `json:"currentPlayerIndex"`
	// RoundFinished distinguishes describing (false) from voting (true)
	// within PhasePlaying.
	RoundFinished bool              `json:"roundFinished"`
	Votes         map[string]string `json:"votes"`
	// LastVotes is the tally of the round that just ended: voter -> target,
	// snapshotted before Votes is cleared so the client can show who voted
	// for whom. Empty until the first tally.
	LastVotes         map[string]string `json:"lastVotes"`
	RolesAcknowledged map[string]bool   `json:"rolesAcknowledged"`
	Winner            string            `json:"winner"` // "" while undecided
	// LastEliminated is nil before the first vote, "" when a vote eliminated
	// nobody (tie or skip), else the eliminated player's name.
	LastEliminated *string `json:"lastEliminated"`
	// Settings is the host's word-pool filter, kept between games. Zero
	// values mean "everything" (see Filter) so rows saved before filters
	// existed behave as before.
	Settings Filter `json:"settings"`

	// Sessions maps resume token -> player. Never included in a View.
	Sessions map[string]string `json:"sessions"`
	// Version increases on every mutation so a client can drop a stale view
	// that arrives around a reconnect.
	Version int64 `json:"version"`
}

func New(id, host string, now time.Time) *Lobby {
	l := &Lobby{
		ID:        id,
		Host:      host,
		Players:   []string{host},
		GamePhase: PhaseLobby,
		CreatedAt: now.UnixMilli(),
		Settings:  DefaultFilter(),
		Sessions:  map[string]string{},
	}
	l.clearGame()
	return l
}

// Normalize fills in nil collections, e.g. after loading older JSON.
func (l *Lobby) Normalize() {
	if l.Players == nil {
		l.Players = []string{}
	}
	if l.Roles == nil {
		l.Roles = map[string]string{}
	}
	if l.AlivePlayers == nil {
		l.AlivePlayers = []string{}
	}
	if l.RoundOrder == nil {
		l.RoundOrder = []string{}
	}
	if l.Votes == nil {
		l.Votes = map[string]string{}
	}
	if l.LastVotes == nil {
		l.LastVotes = map[string]string{}
	}
	if l.RolesAcknowledged == nil {
		l.RolesAcknowledged = map[string]bool{}
	}
	if l.Sessions == nil {
		l.Sessions = map[string]string{}
	}
	// Rows from before filters existed have both categories off, which
	// Validate would reject; they meant "everything".
	if !l.Settings.UseChampions && !l.Settings.UseItems {
		l.Settings = DefaultFilter()
	}
}

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

func ValidatePlayerName(name string) error {
	switch {
	case name == "":
		return invalid("Please enter your name.")
	case len([]rune(name)) > 24:
		return invalid("Names must be 24 characters or fewer.")
	case strings.EqualFold(name, SkipVote):
		return invalid("That name is reserved.")
	}
	return nil
}

// NormalizeLobbyID is the canonical spelling of a code: codes are not case
// sensitive, so "abc12", "Abc12" and "ABC12" all name the same lobby.
func NormalizeLobbyID(id string) string {
	return strings.ToUpper(strings.TrimSpace(id))
}

// ValidateLobbyID checks an already normalized id.
func ValidateLobbyID(id string) error {
	switch {
	case id == "":
		return invalid("Please enter a lobby ID.")
	case len([]rune(id)) > 64:
		return invalid("Lobby IDs must be 64 characters or fewer.")
	case strings.Contains(id, "/"):
		return invalid(`Lobby IDs cannot contain "/".`)
	case id == "." || id == "..":
		return invalid("Invalid lobby ID.")
	}
	return nil
}

// ---------------------------------------------------------------------------
// Lobby
// ---------------------------------------------------------------------------

func (l *Lobby) Join(name string) error {
	if l.GameStarted {
		return ErrInProgress
	}
	if slices.Contains(l.Players, name) {
		return ErrNameTaken
	}
	l.Players = append(l.Players, name)
	return nil
}

// Leave removes a player from every piece of game state they appear in and,
// if that changes the outcome of the game, resolves it. Returns true when the
// host left, which closes the lobby: the caller is expected to delete it.
func (l *Lobby) Leave(name string) (closed bool) {
	if name == l.Host {
		return true
	}
	l.Players = remove(l.Players, name)
	for token, p := range l.Sessions {
		if p == name {
			delete(l.Sessions, token)
		}
	}
	if !l.GameStarted {
		return false
	}

	delete(l.RolesAcknowledged, name)
	delete(l.Votes, name)
	if i := slices.Index(l.RoundOrder, name); i != -1 && i < l.CurrentPlayerIndex {
		l.CurrentPlayerIndex--
	}
	l.AlivePlayers = remove(l.AlivePlayers, name)
	l.RoundOrder = remove(l.RoundOrder, name)

	// The leaver was the last in the round order: the round is over.
	if l.GamePhase == PhasePlaying && l.CurrentPlayerIndex >= len(l.RoundOrder) {
		l.RoundFinished = true
	}
	if w := l.resolveWinner(); l.GamePhase != PhaseGameOver && w != "" {
		l.GamePhase = PhaseGameOver
		l.Winner = w
	}
	return false
}

// ---------------------------------------------------------------------------
// Game flow
// ---------------------------------------------------------------------------

// SetSettings replaces the word-pool filter. Only the host may, and only
// while no game is running: the pool is drawn at Start.
func (l *Lobby) SetSettings(caller string, f Filter) error {
	if caller != l.Host {
		return ErrNotHost
	}
	if l.GameStarted {
		return invalid("Settings can only be changed in the lobby.")
	}
	if err := f.Validate(); err != nil {
		return err
	}
	l.Settings = f
	return nil
}

func (l *Lobby) Start(caller string, rng *rand.Rand, c *Catalog) error {
	if caller != l.Host {
		return ErrNotHost
	}
	if err := l.Settings.Validate(); err != nil {
		return err
	}
	// A double tap on "Start Game" is not an error.
	if l.GameStarted {
		return nil
	}
	if len(l.Players) < MinPlayers {
		return invalid("Need at least 3 players.")
	}

	word, isChampion, err := DrawWord(l.Settings.Normalized(c), rng, c)
	if err != nil {
		return err
	}
	order, undercover := DrawRoles(l.Players, rng)

	l.clearGame()
	l.GameStarted = true
	l.GamePhase = PhaseRevealing
	l.SelectedWord = word.Name
	l.SelectedIcon = word.Icon
	if l.SelectedIcon == "" {
		l.SelectedIcon = DefaultIcon
	}
	l.SelectedIsChampion = isChampion
	l.Players = slices.Clone(order)
	l.AlivePlayers = slices.Clone(order)
	l.RoundOrder = slices.Clone(order)
	for _, p := range order {
		l.Roles[p] = RoleCivilian
		l.RolesAcknowledged[p] = false
	}
	l.Roles[undercover] = RoleUndercover
	return nil
}

func (l *Lobby) Acknowledge(name string) error {
	if l.GamePhase != PhaseRevealing {
		return nil
	}
	if _, ok := l.RolesAcknowledged[name]; !ok {
		return invalid("You are not in this game.")
	}
	l.RolesAcknowledged[name] = true
	return nil
}

// NextPlayer ends the turn of the player at expectedIndex. A stale or
// duplicate call (e.g. a double tap) is ignored; a call from anyone but the
// current player is rejected.
func (l *Lobby) NextPlayer(caller string, expectedIndex int) error {
	if l.GamePhase != PhasePlaying || l.RoundFinished {
		return nil
	}
	if l.CurrentPlayerIndex != expectedIndex {
		return nil
	}
	if l.CurrentPlayerIndex >= len(l.RoundOrder) || l.RoundOrder[l.CurrentPlayerIndex] != caller {
		return invalid("It is not your turn.")
	}
	if l.CurrentPlayerIndex < len(l.RoundOrder)-1 {
		l.CurrentPlayerIndex++
	} else {
		l.RoundFinished = true
	}
	return nil
}

func (l *Lobby) Vote(voter, votedFor string) error {
	if l.GamePhase != PhasePlaying || !l.RoundFinished {
		return invalid("Voting is not open.")
	}
	if !slices.Contains(l.AlivePlayers, voter) {
		return invalid("Only alive players can vote.")
	}
	if votedFor != SkipVote && (votedFor == voter || !slices.Contains(l.AlivePlayers, votedFor)) {
		return invalid("Invalid vote target.")
	}
	l.Votes[voter] = votedFor
	return nil
}

// Reset ends the game (whether finished or aborted by the host) and returns
// the lobby to its pre-game state with the same players.
func (l *Lobby) Reset(caller string) error {
	if caller != l.Host {
		return ErrNotHost
	}
	l.clearGame()
	return nil
}

// Advance applies the transitions that need no player action: role reveal
// -> first round once everyone acknowledged, and voting -> tally once every
// alive player voted. Call it after every mutation. Returns whether the
// lobby changed.
func (l *Lobby) Advance(rng *rand.Rand) bool {
	changed := false
	for l.startRounds() || l.endVotingRound(rng) {
		changed = true
	}
	return changed
}

func (l *Lobby) startRounds() bool {
	if l.GamePhase != PhaseRevealing || len(l.RolesAcknowledged) == 0 {
		return false
	}
	for _, acked := range l.RolesAcknowledged {
		if !acked {
			return false
		}
	}
	l.GamePhase = PhasePlaying
	l.RoundFinished = false
	l.CurrentPlayerIndex = 0
	return true
}

// endVotingRound tallies once every alive player has voted. Ties and skips
// eliminate nobody.
func (l *Lobby) endVotingRound(rng *rand.Rand) bool {
	if l.GamePhase != PhasePlaying || !l.RoundFinished {
		return false
	}
	for _, p := range l.AlivePlayers {
		if _, ok := l.Votes[p]; !ok {
			return false
		}
	}

	voteCount := map[string]int{}
	skipVotes := 0
	for _, voter := range l.AlivePlayers {
		vote := l.Votes[voter]
		switch {
		case vote == SkipVote:
			skipVotes++
		case slices.Contains(l.AlivePlayers, vote):
			voteCount[vote]++
		}
	}
	maxVotes := 0
	for _, n := range voteCount {
		maxVotes = max(maxVotes, n)
	}
	var leaders []string
	for p, n := range voteCount {
		if n == maxVotes {
			leaders = append(leaders, p)
		}
	}
	hasClearWinner := len(leaders) == 1 && maxVotes > 0 && maxVotes > skipVotes

	// Snapshot the ballot before clearing it: the client replays it as the
	// vote-result interstitial, and by now the round is decided so who voted
	// for whom is no longer a secret.
	l.LastVotes = maps.Clone(l.Votes)
	l.Votes = map[string]string{}
	l.RoundFinished = false
	l.CurrentPlayerIndex = 0
	eliminated := ""
	if hasClearWinner {
		eliminated = leaders[0]
		l.AlivePlayers = remove(l.AlivePlayers, eliminated)
	}
	l.LastEliminated = &eliminated

	l.RoundOrder = slices.Clone(l.AlivePlayers)
	rng.Shuffle(len(l.RoundOrder), func(i, j int) {
		l.RoundOrder[i], l.RoundOrder[j] = l.RoundOrder[j], l.RoundOrder[i]
	})

	if w := l.resolveWinner(); w != "" {
		l.GamePhase = PhaseGameOver
		l.Winner = w
	}
	return true
}

// Civilians win when the Undercover is gone; the Undercover wins when they
// are one of the last two players (a 1v1 vote can never remove them).
func (l *Lobby) resolveWinner() string {
	undercoverAlive := false
	for _, p := range l.AlivePlayers {
		if l.Roles[p] == RoleUndercover {
			undercoverAlive = true
			break
		}
	}
	if !undercoverAlive {
		return WinnerCivilians
	}
	if len(l.AlivePlayers) <= 2 {
		return WinnerUndercover
	}
	return ""
}

func (l *Lobby) clearGame() {
	l.GameStarted = false
	l.GamePhase = PhaseLobby
	l.Roles = map[string]string{}
	l.SelectedWord = ""
	l.SelectedIcon = ""
	l.SelectedIsChampion = false
	l.AlivePlayers = []string{}
	l.RoundOrder = []string{}
	l.CurrentPlayerIndex = 0
	l.RoundFinished = false
	l.Votes = map[string]string{}
	l.LastVotes = map[string]string{}
	l.RolesAcknowledged = map[string]bool{}
	l.Winner = ""
	l.LastEliminated = nil
}

func remove(list []string, name string) []string {
	out := make([]string, 0, len(list))
	for _, p := range list {
		if p != name {
			out = append(out, p)
		}
	}
	return out
}

File: lib/models/lobby.dart
import 'package:undercoverleague/models/daily_theme.dart';
import 'package:undercoverleague/models/game_settings.dart';

/// Player roles as the server names them.
class Role {
  static const civilian = 'Civilian';
  static const undercover = 'Undercover';
  static const mrWhite = 'MrWhite';
  static const spectator = 'Spectator';

  static bool isImpostor(String role) => role == undercover || role == mrWhite;

  static String label(String role) => switch (role) {
        mrWhite => 'Mr. White',
        _ => role,
      };
}

/// Game phases as the server names them.
class GamePhase {
  static const lobby = 'lobby';
  static const revealingRoles = 'revealingRoles';
  static const playing = 'playing';
  static const lastGuess = 'lastGuess';
  static const gameOver = 'gameOver';
}

/// One entry of the public clue log (only kept when `clueLog` is on).
class Clue {
  final int round;
  final String player;
  final String text;

  const Clue({required this.round, required this.player, required this.text});

  factory Clue.fromJson(Map<String, dynamic> json) => Clue(
        round: (json['round'] as num?)?.toInt() ?? 0,
        player: json['player'] as String? ?? '',
        text: json['text'] as String? ?? '',
      );

  @override
  bool operator ==(Object other) =>
      other is Clue && other.round == round && other.player == player && other.text == text;

  @override
  int get hashCode => Object.hash(round, player, text);
}

/// The most recent last-breath guess of an eliminated wordless player.
class LastGuess {
  final String player;
  final String word;
  final bool correct;

  const LastGuess({required this.player, required this.word, required this.correct});

  factory LastGuess.fromJson(Map<String, dynamic> json) => LastGuess(
        player: json['player'] as String? ?? '',
        word: json['word'] as String? ?? '',
        correct: json['correct'] as bool? ?? false,
      );

  @override
  bool operator ==(Object other) =>
      other is LastGuess && other.player == player && other.word == word && other.correct == correct;

  @override
  int get hashCode => Object.hash(player, word, correct);
}

/// Per-player counters the server keeps for achievements.
class PlayerStats {
  final int civilianSurvivals;
  final int impostorGames;
  final int games;

  const PlayerStats({this.civilianSurvivals = 0, this.impostorGames = 0, this.games = 0});

  factory PlayerStats.fromJson(Map<String, dynamic> json) => PlayerStats(
        civilianSurvivals: (json['civilianSurvivals'] as num?)?.toInt() ?? 0,
        impostorGames: (json['impostorGames'] as num?)?.toInt() ?? 0,
        games: (json['games'] as num?)?.toInt() ?? 0,
      );

  @override
  bool operator ==(Object other) =>
      other is PlayerStats &&
      other.civilianSurvivals == civilianSurvivals &&
      other.impostorGames == impostorGames &&
      other.games == games;

  @override
  int get hashCode => Object.hash(civilianSurvivals, impostorGames, games);
}

/// One player's view of a lobby, as sent by the server. Secrets are already
/// filtered server-side: only this player's role is included, the word is
/// absent for a wordless player, and [roles] / [selectedWord] are only
/// present once the game is over. [lastVotes] is the exception that is meant
/// to be public: it only ever holds a ballot that has already been tallied.
class Lobby {
  final String id;
  final String host;
  final List<String> players;

  /// Players sitting this game out; a subset of [players].
  final List<String> spectators;
  final bool gameStarted;
  final String gamePhase; // lobby | revealingRoles | playing | lastGuess | gameOver
  final List<String> alivePlayers;
  final List<String> roundOrder;
  final int currentPlayerIndex;
  final bool roundFinished; // false = describing, true = voting

  /// 1-based describing round; 0 in the lobby.
  final int round;

  /// Unix milliseconds when the current turn / vote / guess times out; 0 = no timer.
  final int deadline;
  final Map<String, String> votes;

  /// The ballot of the round that just ended (voter -> target, or 'skip').
  /// Empty until the first tally.
  final Map<String, String> lastVotes;

  /// Every tallied ballot of this game, oldest first.
  final List<Map<String, String>> ballots;
  final Map<String, bool> rolesAcknowledged;

  /// Public clue log of this game (only filled when `clueLog` is on).
  final List<Clue> clues;

  /// Who is making a last guess; '' outside the `lastGuess` phase.
  final String guesser;

  /// The latest last guess of this game; null until the first one.
  final LastGuess? lastGuess;
  final String? winner; // Civilians | Undercover | MrWhite

  /// Why the game ended: eliminated | outnumbered | guess. Null while running.
  final String? winReason;

  /// null before the first vote, '' when a vote eliminated nobody.
  final String? lastEliminated;

  /// Which pack the word was drawn from ('' before a game).
  final String selectedPack;

  final String myRole; // Civilian | Undercover | MrWhite | Spectator
  final String? myWord;
  final String myIcon;

  /// True when [myWord] is a decoy (an Undercover in decoy mode).
  final bool myDecoy;

  final Map<String, String> roles; // empty until game over
  final String? selectedWord; // null until game over

  /// The Undercovers' decoy word; only at game over and only in decoy mode.
  final String? decoyWord;

  /// The host's settings, normalized by the server.
  final GameSettings settings;

  /// Only present in the lobby phase: what [settings] can draw from, which
  /// seasons the sliders may span, the classes, regions, resource buckets
  /// and lanes in the catalog and today's theme. [lanes] is empty when the
  /// server has no play-rate data, and the lane filter is then not offered.
  final PoolSize? poolSize;
  final SeasonRange? seasonRange;
  final List<String> classes;
  final List<String> regions;
  final List<String> resources;
  final List<String> lanes;
  final DailyTheme? dailyTheme;

  /// Lobby-lifetime scoreboard and achievements (player -> ...).
  final Map<String, int> scores;
  final int gamesPlayed;
  final Map<String, List<String>> achievements;
  final Map<String, PlayerStats> stats;

  final Map<String, bool> connected;
  final int version;

  const Lobby({
    required this.id,
    required this.host,
    required this.players,
    this.spectators = const [],
    required this.gameStarted,
    required this.gamePhase,
    required this.alivePlayers,
    required this.roundOrder,
    required this.currentPlayerIndex,
    required this.roundFinished,
    this.round = 0,
    this.deadline = 0,
    required this.votes,
    this.lastVotes = const {},
    this.ballots = const [],
    required this.rolesAcknowledged,
    this.clues = const [],
    this.guesser = '',
    this.lastGuess,
    required this.winner,
    this.winReason,
    required this.lastEliminated,
    String? selectedPack,
    // Older call sites pass the boolean; [selectedPack] wins when both are given.
    bool selectedIsChampion = false,
    required this.myRole,
    required this.myWord,
    required this.myIcon,
    this.myDecoy = false,
    required this.roles,
    required this.selectedWord,
    this.decoyWord,
    this.settings = const GameSettings(),
    this.poolSize,
    this.seasonRange,
    this.classes = const [],
    this.regions = const [],
    this.resources = const [],
    this.lanes = const [],
    this.dailyTheme,
    this.scores = const {},
    this.gamesPlayed = 0,
    this.achievements = const {},
    this.stats = const {},
    required this.connected,
    required this.version,
  }) : selectedPack = selectedPack ?? (selectedIsChampion ? WordPack.champions : '');

  factory Lobby.fromJson(Map<String, dynamic> json) {
    return Lobby(
      id: json['id'] as String? ?? '',
      host: json['host'] as String? ?? '',
      players: _strings(json['players']),
      spectators: _strings(json['spectators']),
      gameStarted: json['gameStarted'] as bool? ?? false,
      gamePhase: json['gamePhase'] as String? ?? GamePhase.lobby,
      alivePlayers: _strings(json['alivePlayers']),
      roundOrder: _strings(json['roundOrder']),
      currentPlayerIndex: _int(json['currentPlayerIndex']),
      roundFinished: json['roundFinished'] as bool? ?? false,
      round: _int(json['round']),
      deadline: _int(json['deadline']),
      votes: _map<String>(json['votes']),
      lastVotes: _map<String>(json['lastVotes']),
      ballots: [for (final b in (json['ballots'] as List?) ?? const []) _map<String>(b)],
      rolesAcknowledged: _map<bool>(json['rolesAcknowledged']),
      clues: [for (final c in (json['clues'] as List?) ?? const []) Clue.fromJson((c as Map).cast())],
      guesser: json['guesser'] as String? ?? '',
      lastGuess: json['lastGuess'] is Map ? LastGuess.fromJson((json['lastGuess'] as Map).cast()) : null,
      winner: json['winner'] as String?,
      winReason: _nonEmpty(json['winReason']),
      lastEliminated: json['lastEliminated'] as String?,
      // Old servers sent a boolean; keep reading it so a mixed deploy works.
      selectedPack: json['selectedPack'] as String? ?? (json['selectedIsChampion'] == true ? WordPack.champions : ''),
      myRole: json['myRole'] as String? ?? Role.spectator,
      myWord: json['myWord'] as String?,
      myIcon: json['myIcon'] as String? ?? 'assets/default_icon.jpg',
      myDecoy: json['myDecoy'] as bool? ?? false,
      roles: _map<String>(json['roles']),
      selectedWord: json['selectedWord'] as String?,
      decoyWord: _nonEmpty(json['decoyWord']),
      settings: json['settings'] is Map ? GameSettings.fromJson((json['settings'] as Map).cast()) : const GameSettings(),
      poolSize: json['poolSize'] is Map ? PoolSize.fromJson((json['poolSize'] as Map).cast()) : null,
      seasonRange: json['seasonRange'] is Map ? SeasonRange.fromJson((json['seasonRange'] as Map).cast()) : null,
      classes: _strings(json['classes']),
      regions: _strings(json['regions']),
      resources: _strings(json['resources']),
      lanes: _strings(json['lanes']),
      dailyTheme: json['dailyTheme'] is Map ? DailyTheme.fromJson((json['dailyTheme'] as Map).cast()) : null,
      scores: {for (final e in _map<Object?>(json['scores']).entries) e.key: _int(e.value)},
      gamesPlayed: _int(json['gamesPlayed']),
      achievements: {for (final e in _map<Object?>(json['achievements']).entries) e.key: _strings(e.value)},
      stats: {
        for (final e in _map<Object?>(json['stats']).entries)
          if (e.value is Map) e.key: PlayerStats.fromJson((e.value as Map).cast()),
      },
      connected: _map<bool>(json['connected']),
      version: _int(json['version']),
    );
  }

  bool get isGameOver => gamePhase == GamePhase.gameOver;
  bool get isLastGuess => gamePhase == GamePhase.lastGuess;

  /// Kept for screens written against the boolean; prefer [selectedPack].
  bool get selectedIsChampion => selectedPack == WordPack.champions;

  String? get currentPlayer =>
      currentPlayerIndex < roundOrder.length ? roundOrder[currentPlayerIndex] : null;

  bool isSpectator(String name) => spectators.contains(name);

  /// Everyone who plays this game: [players] minus the spectators.
  List<String> get activePlayers => players.where((p) => !spectators.contains(p)).toList();

  /// Names of the Undercover(s); only known once the game is over.
  String get undercoverNames =>
      roles.entries.where((e) => e.value == Role.undercover).map((e) => e.key).join(', ');

  /// Names of every impostor (Undercovers and Mr. Whites); only known once
  /// the game is over.
  String get impostorNames =>
      roles.entries.where((e) => Role.isImpostor(e.value)).map((e) => e.key).join(', ');

  /// Time left on the current deadline, null when no timer is running.
  /// Never negative: an overdue deadline reads as zero until the server
  /// moves on.
  Duration? deadlineRemaining(DateTime now) {
    if (deadline == 0) return null;
    final left = deadline - now.millisecondsSinceEpoch;
    return Duration(milliseconds: left < 0 ? 0 : left);
  }

  static List<String> _strings(Object? v) => (v as List?)?.cast<String>() ?? const [];

  static Map<String, T> _map<T>(Object? v) => (v as Map?)?.cast<String, T>() ?? const {};

  static int _int(Object? v) => (v as num?)?.toInt() ?? 0;

  static String? _nonEmpty(Object? v) => v is String && v.isNotEmpty ? v : null;
}
